### PR TITLE
Explicitly clear Gem paths (seems to be required for Ruby 2.3 compat)

### PR DESCRIPTION
**Pull request by alindeman**
I've noticed that, at least with Ruby 2.3.0 and chruby (which I'm obviously trying to avoid having affect anything here), `Gem.path` is already set and cached by the time the code in `env_sanitizer` runs. Only by explicitly clearing the paths via `Gem.clear_paths` can I get any existing GEM_PATH or GEM_HOME out of the path list.

### CODE REVIEW

**Comment by alindeman** [comment]
It's worth noting that `clear_paths` is a bit of a confusing name. All it means is it clears _cached_ paths. From the docs:

> The next time dir or path is requested, the values will be calculated from scratch.

**Comment by alindeman** [comment]
Something is still not quite right with my environment, even after this change. Going to dig deeper before I'm confident in having this PR go through.

**Comment by alindeman** [comment]
Cool, these changes actually do fix an issue for me on Ruby 2.3.0. My earlier comment was caused by an unrelated thing about my environment.

File: lib/appbundler/app.rb [hunk 1]
```diff
@@ -90,7 +90,7 @@ def ruby_relative_path
     end
 
     def shebang
-      "#!#{ruby}\n"
+      "#!#{ruby} --disable-gems\n"
```
Comment (alindeman) [inline lib/appbundler/app.rb]: Disables `rubygems` from being included so we have a chance to clear `GEM_HOME` and `GEM_PATH`. We require it ourselves (`require "rubygems"`) below.

**Comment by alindeman** [comment]
Any thoughts/reviews?

**Comment by jkeiser** [comment]
@alindeman does `clear_paths` actually remove paths from `Gem.path` or does it just clear the cache? @lamont-granquist ^^

**Comment by lamont-granquist** [comment]
clears most of the internal cached state of rubygems:  https://github.com/rubygems/rubygems/blob/d32a94848f50d6be448d862e43f074b820a1d3b6/lib/rubygems.rb#L304-L309

File: lib/appbundler/app.rb [hunk 2]
```diff
@@ -114,7 +114,11 @@ def file_format_comment
     # APPBUNDLER_ALLOW_RVM environment variable to "true". This feature
     # exists to make tests run correctly on travis.ci (which uses rvm).
     def env_sanitizer
-      %Q{ENV["GEM_HOME"] = ENV["GEM_PATH"] = nil unless ENV["APPBUNDLER_ALLOW_RVM"] == "true"}
+      <<-EOS
+ENV["GEM_HOME"] = ENV["GEM_PATH"] = nil unless ENV["APPBUNDLER_ALLOW_RVM"] == "true"
+require "rubygems"
+::Gem.clear_paths if ::Gem.respond_to?(:clear_paths)
```
Comment (lamont-granquist) [inline lib/appbundler/app.rb]: i don't think we care about anything old enough to need to bother with either requiring rubygems or protecting the Gem.clear_paths call.

File: .travis.yml [hunk 3]
```diff
@@ -6,6 +6,7 @@ rvm:
   - 2.0.0
```
Comment (lamont-granquist) [inline .travis.yml]: 2.0 can be removed

File: lib/appbundler/app.rb (hunk 1 above)
Comment (lamont-granquist) [inline lib/appbundler/app.rb]: oh i see, still drinking coffee...

File: lib/appbundler/app.rb (hunk 2 above)
Comment (lamont-granquist) [inline lib/appbundler/app.rb]: still don't have to worry about protecting Gem.clear_paths

**Comment by alindeman** [comment]
> @alindeman does clear_paths actually remove paths from Gem.path or does it just clear the cache? @lamont-granquist ^^

The method name is a bit confusing, but it's clear to me that it only busts _caches_. They'll be rebuilt (with the new values we set) the next time rubygems needs those paths.

**Comment by alindeman** [comment]
@lamont-granquist @jkeiser I think I've addressed your code review comments. Thanks for taking a look.

**Comment by jkeiser** [comment]
@alindeman I'm really sorry it's taken this long to get back to this ... your checkin would have saved us a ton of trouble, and it's clear to me that we need to get this repository onto our triage docket so we don't miss peoples' PRs. @mwrock started making a super similar thing in #29. Let's make sure we don't lose *anything* here, either way (whether we take this verbatim or incorporate there due to some other things being fixed, or whatever).

**Comment by mwrock** [comment]
I have confirmed that @alindeman's changes indeed fix ruby 2.3 lets take this one instead. :+1:

**Comment by jkeiser** [comment]
:+1:

**Comment by lamont-granquist** [comment]
👍

**Pull request merged**